mainmatter/implementation/index.tex

```diff
@@ -1,6 +1,51 @@
 \chapter{Implementation}\label{ch:implementation}
-\begin{chaptermeta}
-Having covered the technical details of both Direct3D and Vulkan, we wish to perform our own tests of the \glspl{API}.
-In this chapter we describe the applications we have developed for the purposed of testing.
-Both how and why they were implemented. 
-\end{chaptermeta}
\ No newline at end of file
+Having now discussed the technical details of the \glspl{API}, this chapter will describe how Direct3D and Vulkan were used to implement the same application. 
+By developing an application, we wish to get familiar with a considerable part of both \glspl{API}, while also testing some of the claims made by their developers.
+
+
+It was decided that the application should be able to render an arbitrary number of textured cubes in 3D space.
+Cubes were chosen, as they are the simplest model to define in 3D space.
+Alternatively, we could have decided to render a more complex model.
+We chose to go with cubes, as this allowed us to keep our rendering system simpler.
+Using a complex model, we would have to write a system to read vertex data from a model file.
+In addition, the process of rendering a cube versus a complex model are not that different as to make the additional effort worthwhile. 
+
+
+We have decided that \gls{GPU} commands should be constructed using multithreading.
+Both \glspl{API} promise that they can be used in the case of \gls{CPU}-bound applications, and we would like to test if this is the case.
+
+As both \glspl{API} expose their render state, rather than keeping it implicit, we want to test how to switch between predefined render pipelines during runtime.
+We will consider how easy it is to work with pipeline objects, and how switching between them affects performance.
+As shaders are defined for each pipeline, this will be done by using different shaders for cubes in the same scene.
+
+ 
+For the application we keep the shaders rather simple.
+This is done as to focus on the use of the \glspl{API} rather than the shader languages.
+
+
+We decided to write the application in C++, as both Vulkan and Direct3D expose \gls{API}headers for use with C and C++.
+While bindings exist for unmanaged languages like C\# and Java, we choose to disregard them as they add a layer of indirection, which negatively affects performance.
+C++ was chosen over C because of our own familiarity with the language. 
+
+Both implementations render their frames to a win32 window.
+We could have used a library like GLFW to create more portable code for handeling windows \cite{GLFW}. 
+Yet, to have more control over how our windows are handled, we decided to use win32 windows directly.  
```

Comment:
Win32 og kan stadig ikke lide brugen af 'yet'

Reply:
Og hvordan giver det os mere kontrol?

Reply:
Og vi valgte det udfra at programmerne skal være samlignlige og Direct3D ikke er understøttet på andet end Windows platformen. Derfor er ekstra lag af indirekthed ikke nødvendig for at skabe platform uafhængighed